### ISSUE TITLE
Sprite Events

### ISSUE THREAD

**Issue by sedabull** [sedabull/Greenhorn#5]
CoffeeScript and JavaScript are event driven languages, and gaming depends very heavily on events. It is therefore only natural that the `Sprite` class should follow some sort of event emitter pattern, so that different handlers can be assigned to different events. The events I'm currently considering are as follows:
* **updating**
    * on('update', callback)
    * on('start', callback)
    * on('stop', callback)
    * on('draw', callback)
* **state changes**
    * on('get:*attr*', callback)
    * on('set:*attr*', callback(to))
    * on('change:*attr*', callback(step))
    * on('*attr*-*cmp*-*value*', callback)
* **boundary collisions**
    * on('off:top', callback)
    * on('off:bottom', callback)
    * on('off:right', callback)
    * on('off:left', callback)
    * on('hit:top', callback)
    * on('hit:bottom', callback)
    * on('hit:right', callback)
    * on('hit:left', callback)
* **non-boundary collisions**
    * on('collisionWith:*other*', callback(other))
    * on('noCollisionWith:*other*', callback(other))
    * on('distanceTo:*other*-*cmp*-*value*', callback(other))
    * on('angleTo:*other*-*cmp*-*value*', callback(other))
* **user input**
    * on('isDown:*key*', callback)
    * on('isUp:*key*', callback)
    * on('mouse:hover', callback)
    * on('mouse:noHover', callback)
* **AniSprite specific**
    * on('play', callback)
    * on('pause', callback)
    * on('stop', callback)

I also intend to review the other classes to determine what events might be appropriate to fire on certain events involving the engine, `Sound` class, `Timer` class etc.

**Comment by sedabull** [comment]
Mouse events are to be fired from the '#gh-canvas' element to all Sprites. The problem with this is that a user will have to remember to check for visibility and a mouse collision in their callback function for these events.
Something like this:
```coffeescript
mySprite.on 'mouse:click', ->
    if @get('visible') and @collidesWith('mouse')
        #do something here
```
It is unfortunate, but I can think of no other way to implement this. The following events will be supported:
* on('mouse:click', callback)
* on('mouse:doubleClick', callback)
* on('mouse:move', callback)
* on('mouse:down', callback)
* on('mouse:up', callback)
* on('mouse:contextMenu', callback)

**Comment by sedabull** [comment]
A user should be allowed to specify event handlers on construction in the config object like so:
```coffeescript
mySprite = new Sprite
    x: 50
    y: 25
    onupdate: ->
        #some code
    ba_all: 'SPRING'
```

**Comment by sedabull** [comment]
The keyboard events have received an upgrade. A list of keys seperated by '-' specifies *disjunction* or an **OR** operation, whereas a list of keys seperated by '+' specifies *conjunction*, or an **AND** operation. For example, an event handler written like this:
```coffeescript
mySprite.on 'isDown:UP-DOWN-RIGHT-LEFT', ->
    #some code here
```
will fire whenever *any* of the arrow keys is down, conversley, something written like this:
```coffeescript
mySprite.on 'isUp:UP+DOWN+RIGHT+LEFT', ->
    #some code here
```
will fire only when *all* of the arrow keys are up.

**Comment by sedabull** [comment]
resolved by #21
